File: LeetCode/34-Find First and Last Position of Element in Sorted Array.py
# 跟上道题一样，还是二分搜索，可以改二分搜索算法，直接找到左边界和右边界。也可以二分搜索找到一个值，然后线性向两端搜索，效率差不多
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        start_index, end_index = -1, -1
        find_index = self.search(nums, target)
        length = len(nums)
        if (find_index != -1):
            start_index = end_index = find_index
            while(start_index - 1 >= 0 and nums[start_index - 1] == target):
                start_index -= 1
            while(end_index + 1 < length and nums[end_index + 1] == target):
                end_index += 1
        return [start_index, end_index]
        
    def search(self, nums, target):
        length = len(nums)
        if (length == 0):
            return -1
        mid_index = len(nums) // 2
        if (nums[mid_index] == target):
            return mid_index
        if (nums[mid_index] < target):
            right_index = self.search(nums[mid_index + 1:], target)
            if (right_index == -1):
                return right_index
            return mid_index + right_index + 1
        
        return self.search(nums[:mid_index], target)

File: LeetCode/34. Find First and Last Position of Element in Sorted Array.cpp
// 基础的二分搜索，先二分找到一个相等值，然后往两边遍历搜索即可，只需要注意边界条件
// 最快答案是两个二分，一个二分找到最左，一个二分找到最右，遇到极端情况确实更佳

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans;
        ans.push_back(-1);
        ans.push_back(-1);
        int start = 0;
        int end = nums.size();
        int mid;
        int i;
        while(start < end) {
            mid = (start + end) / 2;
            if (nums[mid] < target) {
                start = mid + 1;
            }
            else if (nums[mid] > target) {
                end = mid;
            }
            else {
                for (i = mid - 1; i >= 0 && nums[i] == target; --i) {}
                ans[0] = i + 1;
                for (i = mid + 1; i < nums.size() && nums[i] == target; ++i) {}
                ans[1] = i - 1;
                break;
            }
        }
            
        return ans;
    }
};